### ISSUE TITLE
Identify and improve check-checker false positives

### ISSUE THREAD

**Issue by obycode** [hirosystems/clarity-repl#62]
In this issue, I will track commonly seen warnings that could be considered false positives. We can discuss whether or not these are valid warnings and if they are not, then we should implement a solution.

1. Usage of unchecked value after a check on tx-sender
    - "Once I've checked the sender, I trust all of their input"
```clarity
(define-public (set-owner (address principal))
  (begin
    (asserts! (is-eq tx-sender (var-get owner)) (err ERR-NOT-AUTHORIZED))
    (ok (var-set owner address))
  )
)
```

2. Usage of unchecked value after a check on the contract-caller
    - "Once I've checked the caller, I trust all of their input"
```clarity
(define-public (mint-for-dao (amount uint) (recipient principal))
  (begin
    (asserts! (is-eq contract-caller .dao) (err ERR-NOT-AUTHORIZED))
    (ft-mint? our-coin amount recipient)
  )
)
```

**Comment by lgalabru** [comment]
Thanks @obycode, that makes sense to me!
I think a good next improvement could be improving the granularity of the annotations, ie moving from:
`#[allow(unchecked_data)]`
that could lead to undesired effect to something like
`#[check(my-var)]` or `#[untaint(my-var)]`
What do you think?

**Comment by obycode** [comment]
Thanks @lgalabru. Agreed. I have a todo for that already, and now I created issue #64 for this item.

**Comment by lgalabru** [comment]
Oh, I also just noticed the existence of this issue https://github.com/hirosystems/clarity-repl/issues/46

**Comment by lgalabru** [comment]
An interesting case. Not sure if this is a feature or a bug.

<img width="567" alt="Screen Shot 2022-01-05 at 8 33 44 PM" src="https://user-images.githubusercontent.com/87777/148313830-e4c0004a-43f6-4dc0-b0b4-f21a9e23dee7.png">

**Comment by obycode** [comment]
> An interesting case. Not sure if this is a feature or a bug.

Yes, this is a feature. `new-val` is tainted by `step` in the `let`. Maybe we need another message that covers the missing step to make it more clear.

**Comment by lgalabru** [comment]
I like
"Once I've checked the caller, I trust all of their input", 
"Once I've checked the sender, I trust all of their input", 
it makes a lot of sense.

**Comment by obycode** [comment]
Trusted caller/sender are included in 0.25.0. Closing this issue now in favor of more specific issues for each specific case.
